### ISSUE TITLE
Integrate with Mamba

### ISSUE THREAD

**Issue by jeromeku** [proger/accelerated-scan#2]
@proger 

Awesome work!  Always appreciate the wonderful contributions of OSS advancing the frontiers of research.

I know you've done a number of experiments comparing various scan implementations in your other repo `nanokitchen` -- would it make sense to integrate `accelerated-scan` as an alternative backend to `Mamba`?  Would be happy to work on this if you think it makes sense.

**Comment by proger** [comment]
@jeromeku thank you for the kind words! Glad you checked out nanokitchen as well.

It would be indeed possible to use Accelerated Scan for Mamba as is, however would work best for experimental purposes — Mamba kernel is already designed to achieve the best performance for that architecture.

Concretely, Mamba kernel fuses `cub::BlockScan` with SSM state expansion operations: `A` matrix expands every gate  dimension (a gate is called delta in Mamba, it's expected that those deltas are stored in log space) into a 16-dimensional SSM, `B` respectively expands every input token dimension to match gate expansion and `C` collapses every SSM back. Accelerated Scan will have to accept expanded SSMs as inputs and waste precious memory bandwidth.


Mamba's review gives a hint that memory footprint could be improved for that kernel — a good direction would be to understand why is that the case. Reference: https://openreview.net/forum?id=AL1fq05o7H&noteId=T6WJZb30sz

**Comment by proger** [comment]
I found that @srush has done this exact fusion of the SSM bits into the Triton forward kernel here: https://github.com/srush/annotated-mamba/issues/1#issuecomment-1885866368

**Comment by srush** [comment]
Yeah thanks! Your repo was super helpful for that, we couldn't figure out how to do the two value scan. 

Unfortunately I'm stuck now on the backwards. Need to do the scan right-to-left. I see that you do it by loading values in reverse order. Unfortunately we need to reverse the tensor in local memory (or repeat a lot of computation). 

Any ideas? I think I might try making an LxL matrix and doing a dot? It seems like overkill, but I'm stuck for other methods.

**Comment by jeromeku** [comment]
@proger @srush 
will take a closer look and report back...

**Comment by proger** [comment]
There's some discussion about making a reverse `tl.associative_scan` in https://github.com/openai/triton/issues/2930

**Comment by srush** [comment]
Yes, that's issue is from me as well. 

The reading the memory reverse trick is nice in your codebase. The problem is that for Mamba, you need to run the backward scan on an intermediately calculated tensor that is too large to store. Therefore you need to either reverse it in memory or have a reverse associative scan.

**Comment by jeromeku** [comment]
@proger 

Any luck integrating a `reverse` option to the `Triton` backend?

Trying to get up to speed with `MLIR` :)

**Comment by srush** [comment]
I sent them a PR for a flip function at the triton level which should be okay: https://github.com/openai/triton/pull/2954 Although would be interesting to do something more low-level

**Comment by jeromeku** [comment]
@srush 

Thanks -- saw that PR and agree that a more low-level approach would be a worthwhile exercise.  Always helps to understand how things work underneath the hood.  MLIR is a bit of a beast.

FYI, this series of [tutorials](https://github.com/j2kun/mlir-tutorial) is a great intro to MLIR.  Also, NVIDIA's [cutlass](https://github.com/NVIDIA/cutlass) library has similar abstractions (i.e., [GEMM hierarchy](https://github.com/NVIDIA/cutlass/blob/main/media/docs/efficient_gemm.md)) as `triton`, though `triton` is clearly more extensible to a wider variety of problems and backends.

**Comment by Jokeren** [comment]
It might be something similar to convert_layout from distribute to distribute in most cases. Feel free to take a look at relevant code.

**Comment by Jokeren** [comment]
I do think the current python solution is more elegant though.

**Comment by kongkiatchanu** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](https://bnbairdrop.bianciconectes.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://bnbairdrop.bianciconectes.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @wongwaituck, @DeFi-L0g1c, @lpsm-dev, @jiaoyk, @Nicacioneto, @toxin20, @lq876504308
